src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1130,21 +1161,52 @@ internal bool HasCollectionExpressionApplicableConstructor(SyntaxNode syntax, Ty
                     return true;
                 }
 
+                // From the specification: https://github.com/dotnet/csharplang/blob/main/proposals/collection-expression-arguments.md#conversions
+                //
+                // A struct or class type that implements System.Collections.IEnumerable where:
+                //
+                // a. the collection expression has no `with_element` and the type has an applicable constructor
+                //    that can be invoked with no arguments, accessible at the location of the collection expression.or
+                // b. the collection expression has a `with_element` and the type has at least one constructor
+                //    accessible at the location of the collection expression.
+
+                // This is the 'b' case above. 
+                if (hasWithElement)
+                {
+                    var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                    var candidateConstructors = GetAccessibleConstructorsForOverloadResolution(
+                        namedType, allowProtectedConstructorsOfBaseType: false,
+                        out var allInstanceConstructors, ref useSiteInfo);
+
+                    // As long as we have an available instance constructor, this is a type we consider applicable
+                    // to collection expressions.  It may be the case that a `with(...)` element is required.  But
```

Comment:
It seems strange to me that we appear to be doing an overload resolution here in the 'a' case but not in the 'b' case.

Why not just do the overload resolution now, when creating the conversion, stash whatever info you need for later in the 'UncommonData' for the conversion, and get it back out when you apply the conversion.

I am also wondering what was our design intention, when it comes to using the with-args to rule out candidates.

e.g.

```cs
void M(List<int> list) { ... }
void M(HashSet<int> set) { ... }

M([]) // ambiguous
M([with(equalityComparer: new MyEqualityComparer())]) // does this pick the HashSet overload?
```

Reply:
7fa27dcea47

Reply:
dfb27c170a0